### PR TITLE
implemented macros/functions from BSD queue.h. (Fixed #1644)

### LINKED ISSUE #1644: bsd/queue support
Right now we don't have any modules for bsd compatibility, I suggest creating a separate directory for it, like `src/compat/bsd/`.

We need to implement macros/functions from BSD `queue.h`, it can be tested with #1643

### PR DESCRIPTION
**Pull request by subhojit-ghorai**
queue.h has been added. To make queue.h compatible necessary variables and functions have been added to cdefs.h.

### CODE REVIEW

**Comment by Deryugin** [comment]
Please don't create new pull request for every fix, just push more commits to the existing branch.

File: src/compat/posix/include/utmp.h [hunk 1]
```diff
@@ -60,7 +60,7 @@ struct utmp {
 	int32_t ut_addr_v6[4];        /* Internet address of remote
 					host; IPv4 address uses
 					just ut_addr_v6[0] */
-	char __unused[20];            /* Reserved for future use */
+	//char __unused[20];            /* Reserved for future use */
```
Comment (Deryugin) [inline src/compat/posix/include/utmp.h]: Why did you remove this line? Is it related to the quque.h somehow?

Comment (subhojit-ghorai) [reply src/compat/posix/include/utmp.h]: > Please don't create a new pull request for every fix, just push more commits to the existing branch.

Ok, sir. I'll do that

**Comment by subhojit-ghorai** [comment]
> There are lots of unused/empty macros in cdefs.h
> 
> __has_attribute
> __has_extension
> _POSIX_C_SOURCE
> _XOPEN_SOURCE
> __BSD_VISIBLE
> __generic
> __lock_annotate
> 
> And lots of others.
> 
> If you don't need those macros at all, then don't add them.
> If need some of these for the queue.h, please, define it queue.h (or better change queue.h so it doesn't use them at all as I don't see why do we really need those macros).
> 
> Also there are still no any tests, how do you make sure that this code works?

Ill definitely add test commands and push the commit

File: src/compat/posix/include/utmp.h (hunk 1 above)
Comment (subhojit-ghorai) [reply src/compat/posix/include/utmp.h]: > Why did you remove this line? Is it related to the quque.h somehow?

It has been already defined in cdefs.h . So I removed it.

**Comment by subhojit-ghorai** [comment]
Thank you for your reviews. I will make the necessary changes and push the commit

File: src/tests/BSD_queue/queue_test1.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <sys/queue.h>
+#include <embox/test.h>
+
+EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");
+
+struct vacm_view {
+	int				type;
+	//int32_t				status;
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/queue_test1.c]: Rm please

**Comment by anton-bondarev** [comment]
@subhojit-ghorai  Thanks
But It seems that cmd is excess.
Unit test

-  can be named just bsd_queue (not bds_queue1)
-  Use uninitialized variable n (it init in cmd)
-  Use malloc but there are no any free
-  Use static variable `static struct vacm_viewlist vacm_viewlist`

File: src/tests/BSD_queue/queue_test1.c (hunk 2 above)
Comment (subhojit-ghorai) [reply src/tests/BSD_queue/queue_test1.c]: removed

**Comment by subhojit-ghorai** [comment]
> @subhojit-ghorai Thanks
> But It seems that cmd is excess.
> Unit test
> 
> * can be named just bsd_queue (not bds_queue1)
> * Use uninitialized variable n (it init in cmd)
> * Use malloc but there are no any free
> * Use static variable `static struct vacm_viewlist vacm_viewlist`

I have made the requested changes. please review. Thanks

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");
+
+struct test_view {
+	int				type;
+	SLIST_ENTRY(test_view)		vvl;
+};
+
+TEST_CASE("Testing BSD queue functions: ") {
+
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: remove empty line please

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	int n = QUEUE_TEST;                                                       
+
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: rm empty line

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	SLIST_HEAD(test_viewlist, test_view);
+	struct test_viewlist test_viewlist = SLIST_HEAD_INITIALIZER(test_viewlist);
+
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: rm empty line

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	struct test_view *view;
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: split variable declaration and logic

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+	SLIST_INSERT_HEAD(&test_viewlist, view, vvl);
+
+	test_assert_equal(n,1);
+
+	int null_check = ((int)SLIST_EMPTY(&test_viewlist));
+	test_assert_equal(null_check,0);
+
+
+	int num1 = view->type;
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: move declarations to the start of block

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+	struct test_view *view;
+	view = (struct test_view *)malloc(sizeof(*view));
+	memset(view, n, sizeof(*view));
+
+	SLIST_INSERT_HEAD(&test_viewlist, view, vvl);
+
+	test_assert_equal(n,1);
+
+	int null_check = ((int)SLIST_EMPTY(&test_viewlist));
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: move all variable declaration to the start of block

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	test_assert_equal(null_check,0);
+
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: empty line

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	SLIST_HEAD(test_viewlist, test_view);
+	struct test_viewlist test_viewlist = SLIST_HEAD_INITIALIZER(test_viewlist);
+
+	struct test_view *view;
+	view = (struct test_view *)malloc(sizeof(*view));
+	memset(view, n, sizeof(*view));
+
+	SLIST_INSERT_HEAD(&test_viewlist, view, vvl);
+
+	test_assert_equal(n,1);
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: (n,1); -> (n, 1);

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+
+	int num1 = view->type;
+	int num2 = test_viewlist.slh_first->type;
+	test_assert_equal(num1,num2);
+
+
+	SLIST_REMOVE(&test_viewlist,view,test_view,vvl);
+	null_check = ((int)SLIST_EMPTY(&test_viewlist));
+	test_assert_equal(null_check,1);
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: (null_check,1); -> (null_check, 1);

File: src/tests/BSD_queue/BSD_queue_test.c [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+	int null_check = ((int)SLIST_EMPTY(&test_viewlist));
+	test_assert_equal(null_check,0);
+
+
+	int num1 = view->type;
+	int num2 = test_viewlist.slh_first->type;
+	test_assert_equal(num1,num2);
+
+
```
Comment (anton-bondarev) [inline src/tests/BSD_queue/BSD_queue_test.c]: remove empty line

**Comment by anton-bondarev** [comment]
@subhojit-ghorai Please remove  src/cmds/BSD_queue folder from PR

**Comment by subhojit-ghorai** [comment]
> @subhojit-ghorai Please remove src/cmds/BSD_queue folder from PR

Ok. I'll do that.

**Comment by subhojit-ghorai** [comment]
I have made the requested changes. please review. Thanks

**Comment by anton-bondarev** [comment]
> I have made the requested changes. please review. Thanks

Thanks!
When I said about splitting logic and declarations I ment follow
 
```
EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");

struct test_view {
	int				type;
	SLIST_ENTRY(test_view)		vvl;
};

TEST_CASE("Testing BSD queue functions: ") {
	int n = QUEUE_TEST;
	SLIST_HEAD(test_viewlist, test_view);
	struct test_viewlist test_viewlist =       
	SLIST_HEAD_INITIALIZER(test_viewlist);                                           
	struct test_view *view;
	int null_check_before;
	int num1;
	int num2;
	int null_check_after;

	view = (struct test_view *)malloc(sizeof(*view));
	memset(view, n, sizeof(*view));
	
	SLIST_INSERT_HEAD(&test_viewlist, view, vvl);

	null_check_before = ((int)SLIST_EMPTY(&test_viewlist));
	num1 = view->type;

	num2 = test_viewlist.slh_first->type;
	SLIST_REMOVE(&test_viewlist, view, test_view, vvl);
	null_check_after = ((int)SLIST_EMPTY(&test_viewlist));
...
```

**Comment by subhojit-ghorai** [comment]
> > I have made the requested changes. please review. Thanks
> 
> Thanks!
> When I said about splitting logic and declarations I ment follow
> 
> ```
> EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");
> 
> struct test_view {
> 	int				type;
> 	SLIST_ENTRY(test_view)		vvl;
> };
> 
> TEST_CASE("Testing BSD queue functions: ") {
> 	int n = QUEUE_TEST;
> 	SLIST_HEAD(test_viewlist, test_view);
> 	struct test_viewlist test_viewlist =       
> 	SLIST_HEAD_INITIALIZER(test_viewlist);                                           
> 	struct test_view *view;
> 	int null_check_before;
> 	int num1;
> 	int num2;
> 	int null_check_after;
> 
> 	view = (struct test_view *)malloc(sizeof(*view));
> 	memset(view, n, sizeof(*view));
> 	
> 	SLIST_INSERT_HEAD(&test_viewlist, view, vvl);
> 
> 	null_check_before = ((int)SLIST_EMPTY(&test_viewlist));
> 	num1 = view->type;
> 
> 	num2 = test_viewlist.slh_first->type;
> 	SLIST_REMOVE(&test_viewlist, view, test_view, vvl);
> 	null_check_after = ((int)SLIST_EMPTY(&test_viewlist));
> ...
> ```

Thanks for the review. I'll keep it in mind from next time. Should I now make those changes and create a commit?

**Comment by anton-bondarev** [comment]
@subhojit-ghorai 

> Thanks for the review. I'll keep it in mind from next time. Should I now make those changes and create a commit?

You shouldn't make the same changes but you should make your code more elegant!
Please try to understand what your code do.
For example, why do you use null_check_before? Can you check without any variables or use only single var is_empty?
Or What does it mean `SLIST_HEAD(test_viewlist, test_view);` declare struct with some field? Why is it in this location?

**Comment by subhojit-ghorai** [comment]
> @subhojit-ghorai
> 
> > Thanks for the review. I'll keep it in mind from next time. Should I now make those changes and create a commit?
> 
> You shouldn't make the same changes but you should make your code more elegant!
> Please try to understand what your code do.
> For example, why do you use null_check_before? Can you check without any variables or use only single var is_empty?
> Or What does it mean `SLIST_HEAD(test_viewlist, test_view);` declare struct with some field? Why is it in this location?




> @subhojit-ghorai
> 
> > Thanks for the review. I'll keep it in mind from next time. Should I now make those changes and create a commit?
> 
> You shouldn't make the same changes but you should make your code more elegant!
> Please try to understand what your code do.
> For example, why do you use null_check_before? Can you check without any variables or use only single var is_empty?
> Or What does it mean `SLIST_HEAD(test_viewlist, test_view);` declare struct with some field? Why is it in this location?

Thanks a lot for the review. I can definitely write the code without using variables but I thought using them would make it more understandable. 
I did not use a single variable but instead used null_check_before and null_check_after because in between them I am removing the head (SLIST_REMOVE), as all the tests are written at the bottom so changing the value of one single variable before the test would create an error.
SLIST_HEAD(test_viewlist, test_view): this means creating a Singly Linked List whose head is test_viewlist and which holds values of type test_view. It is in that location because this creates the linked list, so it is the first line.

**Comment by subhojit-ghorai** [comment]
I can use simple variable names, for example, 
1) make test_viewlist as head
2)test_view as typeofhead

and I'll add comments for all the steps.

**Comment by subhojit-ghorai** [comment]
Sir, is this fine?

`
EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");
struct test_type {
	int				data;
	SLIST_ENTRY(test_type)		test_field;
};
TEST_CASE("Testing BSD queue functions: ") {
	int n = QUEUE_TEST;
	SLIST_HEAD(head, test_type);
	struct head head =       
	SLIST_HEAD_INITIALIZER(head);                                           
	struct test_type *test_data;
	int null_check_before, null_check_after;

	test_data = (struct test_type *)malloc(sizeof(*test_data));
	memset(test_data, n, sizeof(*test_data));

	SLIST_INSERT_HEAD(&head, test_data, test_field);

	null_check_before = ((int)SLIST_EMPTY(&head));

	SLIST_REMOVE(&head, test_data, test_type, test_field);
	null_check_after = ((int)SLIST_EMPTY(&head));

	test_assert_equal(n, 1);
	test_assert_equal(null_check_before, 0);
	test_assert_equal(test_data->data, head.slh_first->data);	
	test_assert_equal(null_check_after, 1);
	
	free(test_data);
}
`

**Comment by subhojit-ghorai** [comment]
Is this fine?

`EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");
struct test_type {
	int				data;
	SLIST_ENTRY(test_type)		test_field;
};
TEST_CASE("Testing BSD queue functions: ") {
	int n = QUEUE_TEST;
	SLIST_HEAD(head, test_type);
	struct head head =       
	SLIST_HEAD_INITIALIZER(head);                                           
	struct test_type *test_data;
	int null_check_before, null_check_after;

	test_data = (struct test_type *)malloc(sizeof(*test_data));
	memset(test_data, n, sizeof(*test_data));

	SLIST_INSERT_HEAD(&head, test_data, test_field);

	null_check_before = ((int)SLIST_EMPTY(&head));

	SLIST_REMOVE(&head, test_data, test_type, test_field);
	null_check_after = ((int)SLIST_EMPTY(&head));

	test_assert_equal(n, 1);
	test_assert_equal(null_check_before, 0);
	test_assert_equal(test_data->data, head.slh_first->data);	
	test_assert_equal(null_check_after, 1);
	
	free(test_data);
}`

**Comment by subhojit-ghorai** [comment]
```
EMBOX_TEST_SUITE("test Singly Linked List Functions of BSD queue.h");

struct test_type {
	int				data;
	SLIST_ENTRY(test_type)		test_field;
};

TEST_CASE("Testing BSD queue functions: ") {
	int n = QUEUE_TEST;
	SLIST_HEAD(head, test_type);
	struct head head =       
	SLIST_HEAD_INITIALIZER(head);                                           
	struct test_type *test_data;
	int null_check_before, null_check_after;

	test_data = (struct test_type *)malloc(sizeof(*test_data));
	memset(test_data, n, sizeof(*test_data));

	SLIST_INSERT_HEAD(&head, test_data, test_field);

	null_check_before = ((int)SLIST_EMPTY(&head));

	SLIST_REMOVE(&head, test_data, test_type, test_field);
	null_check_after = ((int)SLIST_EMPTY(&head));

	test_assert_equal(n, 1);
	test_assert_equal(null_check_before, 0);
	test_assert_equal(test_data->data, head.slh_first->data);	
	test_assert_equal(null_check_after, 1);
	
	free(test_data);
}
```

**Comment by subhojit-ghorai** [comment]
I have made the requested changes. please review. Thanks

**Comment by Deryugin** [comment]
Test is fine, but I'd change it a little bit.
*  What exactly is `n` variable is used for? What is this `QUEUE_TEST` macro?
```
int n = QUEUE_TEST;
...
test_assert_equal(n, 1);
```

* You can check correctness of list state without additional variables, i.e.
```
	SLIST_INSERT_HEAD(&test_head, test_data, test_field);
	null_check = ((int)SLIST_EMPTY(&test_head));
        ...
        test_assert_equal(null_check, 0);
```
can be reduced just to
```
	SLIST_INSERT_HEAD(&test_head, test_data, test_field);
	test_assert_false(SLIST_EMPTY(&test_head));
```

This way it's easier to read code

**Comment by subhojit-ghorai** [comment]
> Test is fine, but I'd change it a little bit.
> 
> * What exactly is `n` variable is used for? What is this `QUEUE_TEST` macro?
> 
> ```
> int n = QUEUE_TEST;
> ...
> test_assert_equal(n, 1);
> ```
> 
> * You can check correctness of list state without additional variables, i.e.
> 
> ```
> 	SLIST_INSERT_HEAD(&test_head, test_data, test_field);
> 	null_check = ((int)SLIST_EMPTY(&test_head));
>         ...
>         test_assert_equal(null_check, 0);
> ```
> 
> can be reduced just to
> 
> ```
> 	SLIST_INSERT_HEAD(&test_head, test_data, test_field);
> 	test_assert_false(SLIST_EMPTY(&test_head));
> ```
> 
> This way it's easier to read code

Thanks for the review. I created a variable QUEUE_TEST in the queue.h with value 1. It was just for testing purposes.

I will implement the required changes and remove QUEUE_TEST.  thanks.

**Comment by subhojit-ghorai** [comment]
I have made the requested changes. please review. Thanks

**Comment by Deryugin** [comment]
I'm trying to build this test, but it gives me compilation error

```
In file included from ./src/include/embox/test.h:32,
                 from src/tests/BSD_queue/BSD_queue_test.c:12:
src/tests/BSD_queue/BSD_queue_test.c: In function ‘__test_case_at_line_19’:
./src/include/framework/test/assert.h:130:35: error: comparison between pointer and integer [-Werror]
  130 |  __test_assertion_handle((actual) == (expected), \
      |                                   ^~
src/tests/BSD_queue/BSD_queue_test.c:36:2: note: in expansion of macro ‘test_assert_equal’
   36 |  test_assert_equal(SLIST_EMPTY(&test_head), NULL);
```

Fix it, plese

**Comment by subhojit-ghorai** [comment]
> I'm trying to build this test, but it gives me compilation error
> 
> ```
> In file included from ./src/include/embox/test.h:32,
>                  from src/tests/BSD_queue/BSD_queue_test.c:12:
> src/tests/BSD_queue/BSD_queue_test.c: In function ‘__test_case_at_line_19’:
> ./src/include/framework/test/assert.h:130:35: error: comparison between pointer and integer [-Werror]
>   130 |  __test_assertion_handle((actual) == (expected), \
>       |                                   ^~
> src/tests/BSD_queue/BSD_queue_test.c:36:2: note: in expansion of macro ‘test_assert_equal’
>    36 |  test_assert_equal(SLIST_EMPTY(&test_head), NULL);
> ```
> 
> Fix it, please.

Ok. I am working on it.

**Comment by subhojit-ghorai** [comment]
I have made the necessary changes to solve the error. Please review. Thanks.

**Comment by Deryugin** [comment]
Ok, now it looks better :)

Few more issues before it's ready to merge:
* `stdio.h` is not used in test, so shouldn't be included
* Remove blank line at the end of the test file
* Rename package from `embox.test.BSD_queue.BSD_queue_test` to `embox.test.bsd.queue_test`

**Comment by subhojit-ghorai** [comment]
> Ok, now it looks better :)
> 
> Few more issues before it's ready to merge:
> 
> * `stdio.h` is not used in test, so shouldn't be included
> * Remove blank line at the end of the test file
> * Rename package from `embox.test.BSD_queue.BSD_queue_test` to `embox.test.bsd.queue_test`

Thanks for the review. I am working on the issues.

**Comment by subhojit-ghorai** [comment]
I have made the necessary changes. Please review. Thanks.

File: src/tests/bsd/Mybuild [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package embox.test.BSD_queue
```
Comment (Deryugin) [inline src/tests/bsd/Mybuild]: Please, rename package to `embox.test.bsd`, as we generally use lower case names for packages.

Comment (subhojit-ghorai) [reply src/tests/bsd/Mybuild]: yes. I will do it right now.

**Comment by subhojit-ghorai** [comment]
> Ok, now it's good :) Thank you for patience.
> 
> Squash your commits into one (to remove fixup commits) and I'll merge it.

Ok. I'll do it
